File: src/Mesh.js
/**
 * @module mesh ... functions to generate the basic map
 *
 *	a mesh is a triangular tessalation of the map.
 *	a mesh includes:
 *		pts	... the original well spaced points
 *		vor	... Voronoi tesselation of those points
 *		vxs	... <x,y> coordinate of each Voronoi vertex
 *		adj	... list of vertex indices of neighors of each vertex
 *		tris	... list of <x,y> coordinates neighbors of each vertex
 *
 *		edges	... list of [index, index, <x,y>, <x,y>] tupples
 *
 *	O'Leary observed that a map created on a square grid never
 *	loses its regularity, so he wanted to build the map on an
 *	irregular grid.  But he found randomly chosen grids to be
 *	too irregular.  Mesh generation implements his compromise.
 *
 *	1. He starts by generating N completely randomly chosen points.
 *	   But these turn out to be a little to clumpy, so he smoothes
 *	   them out (improves them) by finding the Voronoi polygons
 *	   around those points and using their vertices.
 *
 *	2. He uses those (improved) points as the centers for a
 *	   second Voronoi tesselation.  The edges of those polygons
 *	   are then converted into a triangular grid
 *
 * NOTE: <x,y> coordinates are relative to the center of the map
 */
public class OldMesh {
	
	public MapPoint[] points;	// the points



/**
 * improvePoints: smooth a set of random points
 *
 * @param 	set of <x,y> points
 * @param	number of smoothing iterations
 * @param	extent (range limits)
 * @return	list of smoother <x,y> coordinates
 *
 * each iteration smooths out the distribution of the points
 *	for each point in the set
 *	    generate surrounding Voronoi polygon
 *	    use the centroid of that polygon
 */
function improvePoints(pts, n, extent) {
    n = n || 1;
    extent = extent || defaultExtent;
    for (var i = 0; i < n; i++) {
        pts = voronoi(pts, extent)
            .polygons(pts)
            .map(centroid);
    }
    return pts;
}

/**
 * generateGoodPoints: generate attractive random grid
 *
 * @param	number of points
 * @param	extent (range limits)
 * @return	list of <x,y> coordinates
 *
 * 1. generate a set of random points in the map
 * 2. run one improvement iteration on them
 */
function generateGoodPoints(n, extent) {
    extent = extent || defaultExtent;
    var pts = generatePoints(n, extent);
    pts = pts.sort(function (a, b) {
        return a[0] - b[0];
    });
    return improvePoints(pts, 1, extent);
}

// identify the Voronoi sets associated with a set of points
/**
 * voronoi: compute the Voronoi tesselation for a set or points
 *
 * @param	list of <x,y> coordinates
 * @param	extent (range limits)
 * @param	list of Voronoi regions
 */
function voronoi(pts, extent) {
    extent = extent || defaultExtent;
    var w = extent.width/2;
    var h = extent.height/2;
    return d3.voronoi().extent([[-w, -h], [w, h]])(pts);
}

/**
 * makeMesh - turn a set of well distributed points into a mesh
 *
 * @param	list of <x,y> coordinates
 * @param	extent (size range)
 */
function makeMesh(pts, extent) {
    extent = extent || defaultExtent;

    // compute the Voronoi polygons
    var vor = voronoi(pts, extent);
    var vxs = [];	// vertex locations
    var vxids = {};	// vertex ID #s
    var adj = [];	// adjacent vertices	
    var edges = [];	// list of vertex IDs and positions
    var tris = [];	// coordinates of neighbors of this vertex

    // for each edge of each Voronoi polygon
    for (var i = 0; i < vor.edges.length; i++) {
	// get the two end points of this edge
        var e = vor.edges[i];
        if (e == undefined) continue;

	// lookup (or assign) their vertex IDs
        var e0 = vxids[e[0]];
        if (e0 == undefined) {
            e0 = vxs.length;	
            vxids[e[0]] = e0;
            vxs.push(e[0]);
        }
        var e1 = vxids[e[1]];
        if (e1 == undefined) {
            e1 = vxs.length;
            vxids[e[1]] = e1;
            vxs.push(e[1]);
        }

	// note that each end-point is adjacent to the other
        adj[e0] = adj[e0] || [];
        adj[e0].push(e1);
        adj[e1] = adj[e1] || [];
        adj[e1].push(e0);

	// add indices and coordinates to known edges
        edges.push([e0, e1, e.left, e.right]);

	// note all edges entering the left end point
        tris[e0] = tris[e0] || [];
        if (!tris[e0].includes(e.left)) tris[e0].push(e.left);
        if (e.right && !tris[e0].includes(e.right)) tris[e0].push(e.right);

	// note all edges entering the right end point
        tris[e1] = tris[e1] || [];
        if (!tris[e1].includes(e.left)) tris[e1].push(e.left);
        if (e.right && !tris[e1].includes(e.right)) tris[e1].push(e.right);
    }

    // the new mesh contains all of these things
    var mesh = {
        pts: pts,	// a set of nicely spaced random points
        vor: vor,	// Voronoi tesselation of those points
        vxs: vxs,	// locations of each vertex
        adj: adj,	// indices of neighbors
        tris: tris,	// coordinates of neighbors
        edges: edges,	// the set of all edges
        extent: extent	// the scale 
    }

    /*
     * mesh.map(f) applies f to every vertex in mesh
     */
    mesh.map = function (f) {
        var mapped = vxs.map(f);
        mapped.mesh = mesh;
        return mapped;
    }
    return mesh;
}


/**
 * generateGoodMesh - top level mesh generation
 *
 * @param	number of desired points
 * @param	extent (size limits)
 * @return	mesh
 */
function generateGoodMesh(n, extent) {
    extent = extent || defaultExtent;
    var pts = generateGoodPoints(n, extent);
    return makeMesh(pts, extent);
}

/**
 * isedge - is a point on the map edge
 *
 * @param	mesh
 * @param	index of point of interest
 * @return	true ... point is on the edge
 *
 * In the final (triangular) grid points on the edge have 
 * only two neighbors, while internal points have 3 or more.
 */
function isedge(mesh, i) {
    return (mesh.adj[i].length < 3);
}

// near edge means in the outer 5% of the map
/**
 * isnearedge - is a point near the map edge
 *
 * @param	mesh
 * @param	index of point of interest
 * @return	true ... point is within 5% of edge
 */
function isnearedge(mesh, i) {
    var x = mesh.vxs[i][0];
    var y = mesh.vxs[i][1];
    var w = mesh.extent.width;
    var h = mesh.extent.height;
    return x < -0.45 * w || x > 0.45 * w || y < -0.45 * h || y > 0.45 * h;
}

/**
 * neighbors - neighbors of a vertex
 *
 * @param	mesh
 * @param	index of point of interest
 * @return	list of indices (of neighboring points)
 */
function neighbours(mesh, i) {
    var onbs = mesh.adj[i];
    var nbs = [];
    for (var i = 0; i < onbs.length; i++) {
        nbs.push(onbs[i]);
    }
    return nbs;
}

/**
 * distance - distance between two points
 *
 * @param	mesh
 * @param	index of first point
 * @param	index of second point
 * @return	(positive) distance between them
 */
function distance(mesh, i, j) {
    var p = mesh.vxs[i];
    var q = mesh.vxs[j];
    return Math.sqrt((p[0] - q[0]) * (p[0] - q[0]) + (p[1] - q[1]) * (p[1] - q[1]));
}


/**
 * visualizePoints - plot points on a map
 *	
 * @param	SVG field 
 *		(1000x1000, centered <0,0>)
 * @param	list of <x,y> coordinates
 *		in range <-0.5,-0.5> to <0.5,0.5>
 */
function visualizePoints(svg, pts) {
    // remove all exising circles from the SVG
    var circle = svg.selectAll('circle').data(pts);
    circle.enter()		// HELP
        .append('circle');	// HELP
    circle.exit().remove();	// HELP

    // translate 0-1 coordinates into 1Kx1K coordinaces
    // with radius of 1% of field with
    d3.selectAll('circle')
        .attr('cx', function (d) {return 1000*d[0]})
        .attr('cy', function (d) {return 1000*d[1]})
        .attr('r', 100 / Math.sqrt(pts.length));
}

/**
 * makeD3Path - construct path connecting a set of points
 *	start at first point, draw line to each subsequent point
 *
 * @param	list of <x,y> coordinates
 * @return	string representation of connecting path
 */
function makeD3Path(path) {
    var p = d3.path();
    p.moveTo(1000*path[0][0], 1000*path[0][1]);
    for (var i = 1; i < path.length; i++) {
        p.lineTo(1000*path[i][0], 1000*path[i][1]);
    }
    return p.toString();
}

/**
 * drawPaths - draw line connecting a set of points
 *
 * @param	SVG field
 * @param	class of path to draw
 * @param	list of <x,y> coordinates
 */
function drawPaths(svg, cls, paths) {
    // remove all existing paths from the SVG
    var paths = svg.selectAll('path.' + cls).data(paths)
    paths.enter()
            .append('path')
            .classed(cls, true)
    paths.exit()
            .remove();

    // draw line along the connecting path
    svg.selectAll('path.' + cls)
        .attr('d', makeD3Path);
}
}

File: src/Mesh.java

import java.awt.Color;

import org.rogach.jopenvoronoi.Edge;
import org.rogach.jopenvoronoi.Face;
import org.rogach.jopenvoronoi.HalfEdgeDiagram;
import org.rogach.jopenvoronoi.Point;
import org.rogach.jopenvoronoi.PointSite;
import org.rogach.jopenvoronoi.Vertex;
import org.rogach.jopenvoronoi.VertexType;
import org.rogach.jopenvoronoi.VoronoiDiagram;

/**
 * @module mesh ... functions to generate the basic map
 *
 *         a mesh is a triangular tessalation of the map. a mesh includes:
 *         
 *         O'Leary observed that a map created on a square grid never loses its
 *         regularity, so he wanted to build the map on an irregular grid. But
 *         he found randomly chosen grids to be too irregular. Mesh generation
 *         implements his compromise.
 *
 *         1. He starts by generating N completely randomly chosen points. But
 *         these turn out to be a little to clumpy, so he smoothes them out
 *         (improves them) by finding the Voronoi polygons around those points
 *         and using their vertices.
 *
 *         2. He uses those (improved) points as the centers for a second
 *         Voronoi tesselation, whose vertices become the map points, and
 *         whose edges become a triangular mesh.
 *
 *         NOTE: <x,y> coordinates are relative to the center of the map
 */
public class Mesh {

	private double x_extent;		// arena width
	private double y_extent;		// arena height
	
	private MapPoint[] vertices;	// grid vertices	
	private Path[] edges;			// mesh connections
	
	private Parameters parms;		// global options
	
	/**
	 * create an initial set of points
	 * 
	 * @param number of points
	 * @param x range
	 * @param y range
	 * @param number of improvement iterations
	 */
	public Mesh(int num_points, double x_extent, double y_extent, int improvements) {
		this.x_extent = x_extent;
		this.y_extent = y_extent;
		parms = Parameters.getInstance();

		// create a set of random points
		MapPoint points[] = new MapPoint[num_points];
		for (int i = 0; i < num_points; i++) {
			double x = x_extent * (Math.random() - 0.5);
			double y = y_extent * (Math.random() - 0.5);
			points[i] = new MapPoint(x, y);
		}
		
		// diagnostic display of original points
		PointsDisplay pd = parms.show_points ? new PointsDisplay("Grid Points", 800, 800, Color.BLACK) : null;
		if (pd != null)
			pd.addPoints(points, PointsDisplay.Shape.CIRCLE, Color.GRAY);
			
		// even out the distribution
		for( int i = 0; i < improvements; i++ )
			points = improve(points);
		
		// diagnostic display of improved points
		if (pd != null) {
			pd.addPoints(points, PointsDisplay.Shape.DIAMOND, Color.WHITE);
			pd.repaint();
		}
		
		// create a Voronoi mesh around the improved points
		makeMesh(points);
		if (parms.show_grid) {
			new MeshDisplay("Raw Grid", edges, 800, 800, Color.BLACK, Color.GRAY);
		}
	}

	/**
	 * truncate values outside the extent box to the edge of the box
	 * 
	 * FIX - should interpolate rather than flatten
	 */
	private static double truncate(double value, double extent) {
		if (value < -extent/2)
			return(-extent/2);
		else if (value > extent/2)
			return(extent/2);
		else
			return(value);
	}
	
	/**
	 * is a point within the arena
	 * 
	 * @param OpenVoronoi vertex
	 * @return if that point is within the box
	 * 
	 * 	(necessary because OpenVoronoi generates points outside the box)
	 */
	private boolean inTheBox(Vertex v) {
		if (v.position.x < -x_extent/2 || v.position.x > x_extent/2)
			return false;
		if (v.position.y < -y_extent/2 || v.position.y > y_extent/2)
			return false;
		return true;
	}
	
	/**
	 * even out (the spacing of) a set of random points
	 * 
	 * O'Leary did this by computing the Voronoi polygons surrounding the
	 * chosen points, and then taking the centroids of those polygons.
	 */
	private MapPoint[] improve( MapPoint[] points) {
		MapPoint newPoints[] = new MapPoint[points.length];
		
		// sort the points (left to right)	??? WHY ???
		MapPoint.quickSort(points, 0, points.length-1);

		// create the Voronoi tesselation
		VoronoiDiagram vd = new VoronoiDiagram();
		for (int i = 0; i < points.length; i++) {
			vd.insert_point_site(new Point(points[i].x, points[i].y));
		}
		HalfEdgeDiagram g = vd.get_graph_reference();
		
		// find the centroid of each voronoi polygon
		int i = 0;
		for (Vertex v : g.vertices) {
			// start with the initial point sites
			if (v.type != VertexType.POINTSITE)
				continue;
			
			// ignoring "infinity" points OpenVoronoi added
			if (v.position.x > x_extent / 2 || v.position.x < -x_extent / 2 || v.position.y > y_extent / 2
					|| v.position.y < -y_extent / 2) {
				continue;
			}

			// find the face that owns the point
			Face chosenFace = null;
			for (Face f: g.faces) {
				PointSite p = (PointSite) f.site;
				if (p.position() != v.position)
					continue;
				chosenFace = f;
				break;
			}
			assert(chosenFace != null);
			
			// walk the edges and average their coordinates
			double x_sum = 0;
			double y_sum = 0;
			int numPoints = 0;
			for (Edge e: g.face_edges(chosenFace)) {
				x_sum += truncate(e.source.position.x, x_extent);
				y_sum += truncate(e.source.position.y, y_extent);
				numPoints++;
			}
			newPoints[i++] = new MapPoint(x_sum/numPoints, y_sum/numPoints);
			
			if (parms.debug_level > 2)
				System.out.println("initial point <" + v.position + "> -> <" + newPoints[i-1] + ">");
		}
		return(newPoints);
	}
	
	
	/**
	 * turn a set of points into a mesh
	 * 		compute the Voronoi tesselation
	 * 		for each edge (that is within the box)
	 * 			add each new end to our vertex list
	 * 			add the edge as path in/out of each vertex
	 */
	private void makeMesh( MapPoint[] points ) {
		// compute the Voronoi teselation of the current point set
		VoronoiDiagram vd = new VoronoiDiagram();
		for (int i = 0; i < points.length; i++) {
			vd.insert_point_site(new Point(points[i].x, points[i].y));
		}
		HalfEdgeDiagram g = vd.get_graph_reference();
		
		// allocate hash table to track known vertices
		MapPointHasher pointhash = new MapPointHasher(g.num_vertices(), x_extent, y_extent);
		PathHasher pathhash = new PathHasher(g.num_edges());

		// locate all the vertices and edges
		for( Edge e: g.edges ) {
			// ignore edges that are not entirely within the arena
			Vertex v1 = e.source;
			if (!inTheBox(v1))
				continue;
			Vertex v2 = e.target;
			if (!inTheBox(v2))
				continue;
			
			// assign/get the vertex ID of each end
			MapPoint p1 = pointhash.findPoint(v1.position.x, v1.position.y);
			MapPoint p2 = pointhash.findPoint(v2.position.x, v2.position.y);
			
			// note that each is a neighbor of the other
			p1.addNeighbor(p2);
			p2.addNeighbor(p1);
			
			// and note the path that connects them
			pathhash.findPath(p1, p2);
		} 
		
		// copy out the list of unique Vertices
		vertices = new MapPoint[pointhash.numVertices];
		for(int i = 0; i < pointhash.numVertices; i++)
			vertices[i] = pointhash.vertices[i];
	
		// copy out the list of unique Edges
		edges = new Path[pathhash.numPaths];
		for(int i = 0; i < pathhash.numPaths; i++)
			edges[i] = pathhash.paths[i];
		
		if (parms.debug_level > 0)
			System.out.println(points.length + " points-> " + vertices.length + "/" + g.num_vertices() + 
					" mesh vertices, " + edges.length + "/" + g.num_edges() + " mesh paths");
	}
}
